frame/evm/src/runner/stack.rs

```diff
@@ -296,15 +302,23 @@ where
 					None => 0,
 				};
 
-				let pov_gas = match executor.state().weight_info() {
-					Some(weight_info) => weight_info
+				// Measure actual proof size usage (or get computed proof size)
+				let actual_proof_size = if let Some(measured_proof_size_after) = get_proof_size() {
+					measured_proof_size_after.saturating_sub(measured_proof_size_before)
+				} else {
+					executor
+						.state()
+						.weight_info()
+						.unwrap_or_default()
 						.proof_size_usage
 						.unwrap_or_default()
-						.saturating_mul(T::GasLimitPovSizeRatio::get()),
-					None => 0,
 				};
 
 				// Post execution.
+				let pov_gas = core::cmp::min(
+					actual_proof_size.saturating_mul(T::GasLimitPovSizeRatio::get()),
+					gas_limit,
+				);
```

Comment:
Not sure we should get the `minimum`. Will we not reach a condition where amy process a transaction that consumes more gas than the limit and still pass because in this call we use the `gas_limit`?

```suggestion
				let pov_gas = actual_proof_size.saturating_mul(T::GasLimitPovSizeRatio::get());
```

Reply:
We discussed this a couple days ago with @ahmadkaouk, and we kind of concluded that this is a "safety check" for users to not get overcharged with a gas execution cost they are not willing to pay (over the indicated `gas_limit`). 

> Will we not reach a condition where amy process a transaction that consumes more gas than the limit and still pass because in this call we use the gas_limit?

I think this won't happen, as the actual check for Out Of Gas is being done before this computation, but maybe @ahmadkaouk can confirm.

Reply:
Yes @Agusrodri, it should never happen that a transaction consumes gas (`pov_gas`, `used_gas` or `storage_gas`) more than the `gas_limit` and still pass, otherwise it's a critical bug. I believe this was added as an security check to make sure that users won't be charged more than the limit if this happen (I don't see other reason for why such check is needed !!). And if it's is the case, I agree with @RomarQ that this should happen at the end when we compute the `effective_gas`.

Reply:
If that's the case I don't see any reason to keep this check there to be honest. In my opinion it's better to remove it.

What do you think @ahmadkaouk @RomarQ?

Reply:
Check removed in 1d38ce2b285f1cc7e6515f27348f391735b8a3a7.